### PR TITLE
gh-120317: Lock around global state in the tokenize module

### PR DESCRIPTION
**Pull request by lysnikolaou**
* Issue: gh-120317

### CODE REVIEW

File: Python/Python-tokenize.c
Comment (Fidget-Spinner) [inline Python/Python-tokenize.c]: Instead of this, would `Py_BEGIN_CRITICAL_SECTION(it)` and `Py_END_CRITICAL_SECTION` work? IIRC that locks per-object, and it's the common idiom in the CPython codebase.

Comment (erlend-aasland) [reply Python/Python-tokenize.c]: Correct; you don't need to add a mutex. Just use the object mutex.

<details>

```diff
diff --git a/Python/Python-tokenize.c b/Python/Python-tokenize.c
index ae62c2cfe88..49b72d50b7e 100644
--- a/Python/Python-tokenize.c
+++ b/Python/Python-tokenize.c
@@ -1,6 +1,6 @@
 #include "Python.h"
 #include "errcode.h"
-#include "internal/pycore_lock.h" // PyMutex
+#include "internal/pycore_critical_section.h" // Py_BEGIN_CRITICAL_SECTION
 #include "../Parser/lexer/state.h"
 #include "../Parser/lexer/lexer.h"
 #include "../Parser/tokenizer/tokenizer.h"
@@ -38,10 +38,6 @@ typedef struct
     PyObject *last_line;
     Py_ssize_t last_lineno;
     Py_ssize_t byte_col_offset_diff;
-
-#ifdef Py_GIL_DISABLED
-    PyMutex mutex;
-#endif
 } tokenizeriterobject;
 
 /*[clinic input]
@@ -79,10 +75,6 @@ tokenizeriter_new_impl(PyTypeObject *type, PyObject *readline,
     }
     self->done = 0;
 
-#ifdef Py_GIL_DISABLED
-    self->mutex = (PyMutex) {_Py_UNLOCKED};
-#endif
-
     self->last_line = NULL;
     self->byte_col_offset_diff = 0;
     self->last_lineno = 0;
@@ -191,13 +183,10 @@ tokenizeriter_next(tokenizeriterobject *it)
     struct token token;
     _PyToken_Init(&token);
 
-#ifdef Py_GIL_DISABLED
-    PyMutex_Lock(&it->mutex);
-#endif
-    int type = _PyTokenizer_Get(it->tok, &token);
-#ifdef Py_GIL_DISABLED
-    PyMutex_Unlock(&it->mutex);
-#endif
+    int type;
+    Py_BEGIN_CRITICAL_SECTION(it);
+    type = _PyTokenizer_Get(it->tok, &token);
+    Py_END_CRITICAL_SECTION();
 
     if (type == ERRORTOKEN) {
         if(!PyErr_Occurred()) {
```

</details>

Comment (lysnikolaou) [reply Python/Python-tokenize.c]: Thanks both! Fixed.

**Comment by Fidget-Spinner** [comment]
This LGTM. Can you add the reproducer to `test.test_free_threading` please? I'll approve after!

**Comment by lysnikolaou** [comment]
@Fidget-Spinner I added more locking around global state. Reviewing each commit seprately might make it easier.

File: Python/Python-tokenize.c
Comment (Fidget-Spinner) [inline Python/Python-tokenize.c:228]: GHA is warning that byte_offset may be uninitialized at this point.

File: Python/Python-tokenize.c
Comment (erlend-aasland) [inline Python/Python-tokenize.c]:
```suggestion
    }
    else {
```

File: Python/Python-tokenize.c
Comment (erlend-aasland) [inline Python/Python-tokenize.c:228]: GitHub complains about `byte_offset` being uninitialised here; this should be fixed before merge, because it'll appear in other PRs as well.

File: Python/Python-tokenize.c
Comment (erlend-aasland) [inline Python/Python-tokenize.c]:
```suggestion
        }
        else {
```

File: Python/Python-tokenize.c
Comment (Fidget-Spinner) [inline Python/Python-tokenize.c:186]: You might want to add a macro that asserts `it` is locked See `_Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED` or `ASSERT_DICT_LOCKED` for a macro redefinition of that.

File: Python/Python-tokenize.c
Comment (Fidget-Spinner) [inline Python/Python-tokenize.c]: Same here about asserting things are locked.

File: Python/Python-tokenize.c
Comment (Fidget-Spinner) [inline Python/Python-tokenize.c:343]: You could change this to an atomic, which might be faster than locking? See the `FT_ATOMIC_STORE_X` macros.

**Review by Fidget-Spinner — COMMENTED** [review]
This looks like a refactor, so good so far.

**Comment by lysnikolaou** [comment]
Feedback addressed.

@Fidget-Spinner I used `_Py_atomic_store_int` instead of the macro. Is that okay?

File: Python/Python-tokenize.c
Comment (Fidget-Spinner) [inline Python/Python-tokenize.c]: Might want to guard with an `ifdef Py_GIL_DISABLED` and keep the original version around if it's perf sensitive.

File: Python/Python-tokenize.c
Comment (Fidget-Spinner) [inline Python/Python-tokenize.c]: Same here

**Review by Fidget-Spinner — APPROVED** [review]
LGTM. Feel free to ignore the following comments if this part of the tokenizer is not perf sensitive.

File: Lib/test/test_free_threading/test_tokenize.py
Comment (Fidget-Spinner) [inline Lib/test/test_free_threading/test_tokenize.py:12]: For this to pass on WASI, you need to `from test.support import threading_helper` and decorate the class with `@threading_helper.requires_working_threading()`

**Comment by lysnikolaou** [comment]
All feedback addressed. Thanks a lot for the reviews and the help @Fidget-Spinner and @erlend-aasland!

**Review by Fidget-Spinner — APPROVED** [review]

File: Python/Python-tokenize.c
Comment (pablogsal) [inline Python/Python-tokenize.c]: Nit: feel free to ignore

```suggestion
    PyObject *line = it->last_line;
    if (it->tok->lineno != it->last_lineno) {
        // Line has changed since last token, so we fetch the new line and cache it
        // in the iter object.
        Py_XDECREF(it->last_line);
        line = PyUnicode_DecodeUTF8(line_start, size, "replace");
        it->last_line = line;
        if (it->tok->lineno != it->last_end_lineno) {
            it->byte_col_offset_diff = 0;
        }
    }
    return line;
}
```

**Review by pablogsal — APPROVED** [review]
LGTM

File: Lib/test/test_free_threading/test_tokenize.py
Comment (pablogsal) [inline Lib/test/test_free_threading/test_tokenize.py:12]: Tested this locally: it crashes spectacularly without the fix :)

File: Python/Python-tokenize.c
Comment (pablogsal) [inline Python/Python-tokenize.c]: Don't we need to protect the read? Otherwise we may get weird partial reads no?

**Comment by pablogsal** [comment]
Actually, one question here: shouldn't we be protecting all reads of the `it` structure that are not under critical sections? Otherwise we are getting some partial reads no? I am missing something?

**Comment by pablogsal** [comment]
For example, this read is not protected, no?:

```
    Py_ssize_t lineno = ISSTRINGLIT(type) ? it->tok->first_lineno : it->tok->lineno;
     Py_ssize_t end_lineno = it->tok->lineno;
```

**Comment by lysnikolaou** [comment]
After some offline discussion with @pablogsal, we decided it makes more sense to just lock around all of `tokenizeriter_next`.

File: Python/Python-tokenize.c
Comment (Fidget-Spinner) [inline Python/Python-tokenize.c:246]: I don't think you can do this. In the exit case, it will not unlock the mutex.

Perhaps rename `tokenizeriter_next` to `tokernizeriter_next_unlocked` and then create a new wrapper function `tokenizeriter_next` that just does
```
static PyObject *
tokenizeriter_next(tokenizeriterobject *it) 
{
    PyObject *res;
    Py_BEGIN_CRITICAL_SECTION(it);
    res = tokernizeriter_next_unlocked(it);
    Py_END_CRITICAL_SECTION()
    return res;
}
```

Comment (lysnikolaou) [reply Python/Python-tokenize.c:246]: Why not? The `Py_END_CRITICAL_SECTION` call is after the `exit`, so it'll be invoked ther as well, right?

Comment (Fidget-Spinner) [reply Python/Python-tokenize.c:246]: Wait I think I misread. That should be okay then. Sorry!

File: Lib/test/test_free_threading/test_tokenize.py
Comment (colesbury) [inline Lib/test/test_free_threading/test_tokenize.py]: Can we make this test faster? It's going to run on every CI job. I think we should aim for 0.1 seconds or less for these sorts of tests. (It's currently ~3 seconds).

Comment (lysnikolaou) [reply Lib/test/test_free_threading/test_tokenize.py]: Changed the amount of time each thread sleeps to 0.03 secs, which means that the whole test now takes about 0.1 seconds. Also verified that the test still fails without the fix and succeeds with it.

**Comment by lysnikolaou** [comment]
@pablogsal Can do a final review here when you have some spare cycles?

**Review by pablogsal — APPROVED** [review]
Reviewed the PR and the comments and it LGTM

I also ran the tests locally in free threaded mode to confirm

**Review by erlend-aasland — APPROVED** [review]

**Pull request merged**
